src/main/java/com/conveyal/analysis/controllers/BundleController.java

```diff
@@ -112,140 +115,139 @@ private Bundle create (Request req, Response res) {
                 bundle.north = bundleWithFeed.north;
                 bundle.south = bundleWithFeed.south;
                 bundle.east = bundleWithFeed.east;
                 bundle.west = bundleWithFeed.west;
                 bundle.serviceEnd = bundleWithFeed.serviceEnd;
                 bundle.serviceStart = bundleWithFeed.serviceStart;
                 bundle.feeds = bundleWithFeed.feeds;
-                bundle.feedsComplete = bundleWithFeed.feedsComplete;
-                bundle.totalFeeds = bundleWithFeed.totalFeeds;
+            } else {
+                bundle.serviceStart = LocalDate.MAX;
+                bundle.serviceEnd = LocalDate.MIN;
+                bundle.feeds = new ArrayList<>();
             }
-        } catch (Exception e) {
+        } catch (UnsupportedEncodingException e) {
             throw AnalysisServerException.badRequest(ExceptionUtils.asString(e));
         }
 
         // Set `createdBy` and `accessGroup`
-        bundle.accessGroup = req.attribute("accessGroup");
-        bundle.createdBy = req.attribute("email");
+        bundle.accessGroup = user.accessGroup;
+        bundle.createdBy = user.email;
 
         Persistence.bundles.create(bundle);
 
         // Process OSM first, then each feed sequentially. Asynchronous so we can respond to the HTTP API call.
-        UserPermissions userPermissions = req.attribute(USER_PERMISSIONS_ATTRIBUTE);
-        taskScheduler.enqueue(Task.forUser(userPermissions)
-            .withDescription("Process OSM and GTFS")
-            .setHeavy(true)
+        final Task task = taskScheduler.newTaskForUser(user)
+            .withTag("category", "bundles")
+            .withTag("title", "Process OSM and GTFS for " + bundle.name)
+            .withTag("bundleId", bundle._id)
+            .withTag("regionId", bundle.regionId)
             .withTotalWorkUnits(
+                1 + // Initial creation
+                1 + // write manifest to cache
                 ((bundle.osmId == null) ? 1: 0) +
                 ((bundle.feedGroupId == null) ? files.get("feedGroup").size() : 0)
             )
-            .withAction(progressListener -> {
-              try {
-                if (bundle.osmId == null) {
-                    // Process uploaded OSM.
-                    bundle.status = Bundle.Status.PROCESSING_OSM;
-                    bundle.osmId = new ObjectId().toString();
-                    Persistence.bundles.modifiyWithoutUpdatingLock(bundle);
-
-                    DiskFileItem fi = (DiskFileItem) files.get("osm").get(0);
-                    OSM osm = new OSM(null);
-                    osm.intersectionDetection = true;
-                    osm.readPbf(fi.getInputStream());
-
-                    fileStorage.moveIntoStorage(osmCache.getKey(bundle.osmId), fi.getStoreLocation());
-                    progressListener.increment();
-                }
+            .withLogEntry("Bundle creation started");
+
+        task.withAction(p -> {
```

Comment:
Yes, it was only replaced with `p` because it went unused.

Reply:
Ah, now I see what you meant by "using the task directly" instead of via the progressListener interface and parameter. I didn't catch that yesterday because it's so different from how I expected this to work, but as always I guess it's important to be able to articulate what I intended and why. 

These methods executed in the background are always provided a progress callback as a declared argument. First, this serves as a contract that they are expected (required) to report progress, and ensures their implementer is aware they need to behave as background tasks. Second, and most importantly, this allows classes and (named) non-closure methods to report progress. I think of these as the primary style, with anonymous closures just being a convenient shortcut to prototype new ideas. Using closures as the mechanism to supply the progress reporting callbacks to the background action would block us from using what I consider the "normal" structured or encapsulated approach. 

My intent with TaskScheduler is to build a general purpose component we can rely on and heavily reuse over the long term, which is predictable, stable, and easy to reason about under concurrent use. I never really considered heavy or exclusive use of anonymous functions and closures to communicate with this kind of component, especially since it can have a well defined and simple interface.

The progress listener would not have to be passed in to the method. There are several other ways it could be made available to the method's scope. But passing it as a parameter does ensure it's available and insist on its use, which I like.

I think the difference in perspective comes from the very heavy use of anonymous functions and closures in Javascript. To my mind using closures as an interface between components is a bit like String-String maps, maybe useful while rapidly iterating on a prototype, but not appropriate for a more stable system.